packages/@lwc/ssr-runtime/src/index.ts

```diff
@@ -172,20 +172,29 @@ export class LightningElement implements PropsAvailableAtConstruction {
         return (this.__classList = new ClassList(this));
     }
 
-    getAttribute(attrName: string): string | null {
-        return this.__attrs[attrName] ?? null;
+    #setAttribute(attrName: string, attrValue: string | null): void {
+        this.__attrs[attrName] = attrValue;
     }
 
-    setAttribute(attrName: string, value: string | null): void {
-        this.__attrs[attrName] = String(value);
+    setAttribute(attrName: string, attrValue: unknown): void {
+        this.#setAttribute(attrName, String(attrValue));
     }
 
-    hasAttribute(attrName: string): boolean {
-        return Boolean(this.__attrs && attrName in this.__attrs);
+    getAttribute(attrName: unknown): string | null {
+        if (this.hasAttribute(attrName)) {
+            return this.__attrs[attrName as string];
+        }
+        return null;
+    }
+
+    hasAttribute(attrName: unknown): boolean {
+        return typeof attrName === 'string' && typeof this.__attrs[attrName] === 'string';
     }
 
     removeAttribute(attrName: string): void {
-        this.__attrs[attrName] = null;
+        if (this.hasAttribute(attrName)) {
+            this.#setAttribute(attrName, null);
```

Comment:
```suggestion
            // `__attrs` uses getters/setters, which we need to preserve in case the attribute is set again,
            // so we use `null` to represent the "removed" state
            this.#setAttribute(attrName, null);
```
Might also want to add this as a JSDoc to `#setAttribute` (and maybe elsewhere?).